pkg/didcomm/crypto/jwe/authcrypt/encrypt.go

```diff
@@ -0,0 +1,368 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package authcrypt
+
+import (
+	"crypto"
+	"crypto/cipher"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"sort"
+	"strings"
+
+	josecipher "github.com/square/go-jose/v3/cipher"
+	"golang.org/x/crypto/blake2b"
+	"golang.org/x/crypto/curve25519"
+	"golang.org/x/crypto/nacl/box"
+	errors "golang.org/x/xerrors"
+
+	"github.com/btcsuite/btcutil/base58"
+	"golang.org/x/crypto/chacha20poly1305"
+	"golang.org/x/crypto/poly1305"
+)
+
+// This package deals with Authcrypt encryption for Packing/Unpacking DID Comm exchange
+// Using Chacha20Poly1035 encryption/authentication
+
+// C20P Chacha20Poly1035 algorithm
+const C20P = "C20P"
+
+// XC20P XChacha20Poly1035 algorithm
+const XC20P = "XC20P"
+
+// RandReader is a cryptographically secure random number generator.
+var RandReader = rand.Reader
+
+type keyPair struct {
+	priv *[chacha20poly1305.KeySize]byte
+	pub  *[chacha20poly1305.KeySize]byte
+}
+
+// Crypter represents an Authcrypt Encrypter (Decrypter) that outputs/reads JWE envelopes
+type Crypter struct {
+	sender     keyPair
+	recipients []*[chacha20poly1305.KeySize]byte
+	alg        string
+	nonceSize  int
+}
+
+// Envelope represents a JWE envelope as per the Aries Encryption envelop specs
+type Envelope struct {
+	Protected  string      `json:"protected,omitempty"`
+	Recipients []Recipient `json:"recipients,omitempty"`
+	Aad        string      `json:"aad,omitempty"`
+	Iv         string      `json:"iv,omitempty"`
+	Tag        string      `json:"tag,omitempty"`
+	CipherText string      `json:"ciphertext,omitempty"`
+}
+
+// jweHeaders are the Protected JWE headers in a map format
+type jweHeaders map[string]string
+
+// Recipient is a recipient of an envelop including the shared encryption key
+type Recipient struct {
+	EncryptedKey string           `json:"encrypted_key,omitempty"`
+	Header       RecipientHeaders `json:"header,omitempty"`
+}
+
+// RecipientHeaders are the recipient headers
+type RecipientHeaders struct {
+	Apu string `json:"apu,omitempty"`
+	Iv  string `json:"iv,omitempty"`
+	Tag string `json:"tag,omitempty"`
+	Kid string `json:"kid,omitempty"`
+	Oid string `json:"oid,omitempty"`
+}
+
+// New will create an encrypter instance to 'AuthCrypt' payloads for the given sender and recipients arguments
+// and the encryption alg argument. Possible algorithms supported are:
+// C20P (chacha20-poly1035 ietf)
+// XC20P (xchacha20-poly1035 ietf)
+// The returned crypter contains all the information required to encrypt payloads.
+func New(sender keyPair, recipients []*[chacha20poly1305.KeySize]byte, alg string) (*Crypter, error) {
```

Comment:
Recommending creating a type for alg for better documentation.

(also to conform better with go-jose: https://github.com/square/go-jose/blob/8bad6148bd0a8da57f9840693f6f611ac4483d09/shared.go#L36)

Reply:
should I reference go-jose's type or create a new internal type?

Reply:
I had thought a new type.

Reply:
I was thinking to reference go-jose as they already have a type for this.

I'll create a new type then..

Reply:
done